### ISSUE TITLE
address dart2js code-size regression when enabling cfe constants

### ISSUE THREAD

**Issue by sigmundch** [dart-lang/sdk#37215]
It appears that turning on CFE constants also affects the SSA inlining decisions and as a result  some internal apps get bigger (~ 0.7%).

We need to investigate this and see if it is a small fix (e.g. tweak heuristics that are now affected by the flag change), but the impact is large enough that I'd consider this a blocker before we turn on the flag.

/cc @rakudrama @johnniwinther @vsmenon

**Comment by vsmenon** [comment]
@fishythefish - can you take a look?  Stephen can you give you context.

**Comment by vsmenon** [comment]
Ping - any update on this?  We'd like to enable the flag next week.

**Comment by fishythefish** [comment]
https://dart-review.googlesource.com/c/sdk/+/106727 seems to largely mitigate the issue. There are still some differences in the compiled code, but I need to investigate further to see if they're just differences in constant folding.

**Comment by fishythefish** [comment]
Okay, I've finished digging into what our inliner is seeing, and here's the writeup:

Our SSA builder kernel has an `InlineWeeder` class which walks the AST and tracks certain metrics which we use in our inlining heuristics. The primary metric is node counts: we have a regular node count roughly tracking the number of AST nodes we've seen (although we sometimes adjust for nodes that don't correspond to any actual code output) and a reductive node count, which tracks the subset of nodes whose inlining means a strict size reduction. This system assumes that the size of an AST subtree is essentially proportional to the code we produce for it, but this is not always the case.
_(Below, I'll use the convention that node counts are written as `<reductive>/<regular>`.)_ 

Secondarily, we also track the number of calls. The reductive inliner will refuse to inline a subtree containing more than one call.

With `constant-update-2018` enabled, the new `ConstantExpression` node now appears in the AST. Because we didn't override `visitConstantExpression`, the default behavior is to assign it a node count of 2/2. (1/1 comes from the default when we visit the `ConstantExpression` itself, and 1/1 comes from the default when we recursively visit the `DartType` it contains.) The majority of `ConstantExpression`s replace primitive literal nodes and `StaticGet`s, which all have node counts of 1/1. [This CL (patchset 5)](https://dart-review.googlesource.com/c/sdk/+/106727/5) overrides `visitConstantExpression` to simply increment the node count by 1/1. **This alone reduces code bloat from `constant-update-2018` to approximately 0.18%.**

The remaining differences are the following:
* The CFE reduces the number of `VariableDeclaration`s overall. (Each one formerly had a node count of 1/3.) This is a good thing, and we don't need to do anything to adjust for this, although we may inline more aggressively as a result.
* Constant `ListLiteral`s turn into `ConstantExpression`s. These now have a reduced node count since we no longer visit their type arguments. Ditto for `SetLiteral`s and `MapLiteral`s. This should result in more aggressive inlining.
* Constant `StaticGet`s turn into `ConstantExpression`s. In this case, we fail to increment the call count. This may result in more aggressive inlining, depending on other calls.
* Constant `ConstructorInvocation`s and `StaticInvocation`s turn into `ConstantExpression`s. These have a significantly reduced node count because we no longer visit their argument lists. We also fail to mark these as calls. These should be more aggressively inlined as a result.

Of these differences, former `StaticGet`s are the majority by far. I'll do some experimenting tomorrow to see if I can reduce the code bloat further by adjusting for these.

**Comment by fishythefish** [comment]
### tl;dr

After some more experimenting, I think we should accept the **0.18%** code bloat for now, since it looks like the pre-`constant-update-2018` code was doing the wrong thing. I'll mark this as non-blocking, but I'll leave the issue open and continue to investigate.

---

### Details

When dart2js determines something to be a constant value, we store it in the constant pool and all future accesses are via a constant pool reference. This means that for the purposes of inlining, all constants should have a node count of 1/1. However, the old inlining code generally did not distinguish constants from non-constants and consequently walked AST subtrees for all values. This meant that non-primitive constants were incorrectly given higher node counts.

We did specifically check for constant `ConstructorInvocation`s and `StaticInvocation`s, but it looks like we did the wrong thing here as well: we stopped incrementing the reductive node count for their subtrees, and then proceeded to recursively visit the children and increment the regular node count as usual.

This means that the old code was probably inlining less aggressively than it should have been overall, although the reductive inliner may have been too aggressive when it came to constant `ConstructorInvocation`s and `StaticInvocation`s.

---

### Current Issues

The new code is inlining certain constructor and method calls more aggressively than the old code (and more aggressively than we want). This apparently occurs even when no `ConstantExpression`s or constant pool references are involved. I need to investigate further why `constant-update-2018` is causing this change.

---

**Note**: I did figure out a way to get the code bloat down to about **0.06%** - by having `defaultConstantReference` just visit the `Constant` normally, we can mostly recover the old, incorrect behavior of walking the full subtrees of constant values. However, I don't think we should continue pursuing parity in code size at the cost of persisting poor inlining decisions.

**Comment by fishythefish** [comment]
Deprioritizing this issue for now.

There are several ideas we have for refactoring the inliner to make smarter decisions, but all of those are larger projects than we want to undertake right now. One quick thing we can do to reduce code size further is to tweak our inlining thresholds:
https://github.com/dart-lang/sdk/blob/d563bbba274c36e1b6fa0c48caab798f78e1ed2c/pkg/compiler/lib/src/ssa/builder_kernel.dart#L6766-L6771

These are essentially magic numbers which determine how small a function's node count has to be in order to be inlined. Because `constant-update-2018` lowers node counts slightly, a handful of constructors/methods whose node counts were above the inlining thresholds are now below them. These functions are consequently inlined more aggressively, causing code bloat. If we lower the thresholds correspondingly (e.g. from `15` and `34` to `14` and `32`, respectively), this eliminates the code bloat entirely. This does, however, now cause some *other* functions to be inlined less aggressively.

In general, lowering inlining thresholds causes less inlining, resulting in lower code size but potentially poorer performance, so fine tuning this will require further iteration. We should wait for `constant-update-2018` to be enabled internally before doing this so that our customers don't see sudden fluctuations in these metrics.

**Comment by joshualitt** [comment]
Even though there is still remaining work here, I'm closing this issue due to fishythefish's updates.
